xapian-core/weight/tfidfweight.cc

```diff
@@ -153,18 +188,29 @@ TfIdfWeight::get_idfn(Xapian::doccount termfreq, char c) const
 	    return (1.0 / termfreq);
 	case 's':
 	    return pow(log(N / termfreq), 2.0);
+	case 'P':
+	    return (log((N+1) / termfreq));
 	default:
 	    AssertEq(c, 't');
 	    return (log(N / termfreq));
     }
 }
 
 double
-TfIdfWeight::get_wtn(double wt, char c) const
+TfIdfWeight::get_wtn(Xapian::termcount doclen, double wt, char c) const
 {
-    (void)c;
-    AssertEq(c, 'n');
-    return wt;
+    double normlen;
```

Comment:
You should declare variables as locally as possible, which would be within the block for the `'P'` case.